src/elexclarity/formatters/results.py

```diff
@@ -101,7 +128,58 @@ def format_top_level_counts(self, choices):
 
         return counts
 
-    def format_race(self, contest, election_date, election_type, timestamp=None, level=None, county_id=None, **kwargs):
+    def _get_precinct_fully_reporting_statuses_via_percent_reporting(self, election):
+        """
+        Constructs a mapping from precinct IDs to boolean flags representing
+        whether or not the given precinct is fully reporting. This method does so by looking
+        at the VoterTurnout from the ElectionResults tag in the precinct details xml
+        and checking if ``percentReporting`` is 4. This means (we think) that all 4 vote
+        types (early/absentee/day of in person/provisional) have been reported for the
+        given precinct.
+        """
+        subunit_fully_reporting_statuses = {}
+        precincts = get_list(election["VoterTurnout"]["Precincts"]["Precinct"])
+        for precinct in precincts:
+            subunit_id = self.get_precinct_id(precinct["name"])
+            is_precinct_fully_reporting = (precinct.get("percentReporting") == str(self.NUM_VOTE_TYPES))
+            subunit_fully_reporting_statuses[subunit_id] = is_precinct_fully_reporting
+
+        return subunit_fully_reporting_statuses
+
+    def _get_precinct_fully_reporting_statuses_via_vote_types(self, contest):
+        """
+        Constructs a mapping from precinct IDs to boolean flags representing
+        whether or not the given precinct is fully reporting. This method does so by looking
+        at the VoterTurnout from the ElectionResults tag in the precinct details xml
+        and checking if ``percentReporting`` is 4. This means (we think) that all 4 vote
+        types (early/absentee/day of in person/provisional) have been reported for the
+        given precinct.
+        """
+        choices = self._get_valid_contest_choices(contest)
+        subunit_fully_reporting_statuses = {}
+
+        vote_types_by_subunit = self.get_vote_totals_by_vote_types(choices, "precinct")
+        for subunit_id, vote_type_totals in vote_types_by_subunit.items():
+            is_precinct_fully_reporting = True
+            for vote_type, vote_total in vote_type_totals.items():
+                if vote_total == 0 and vote_type not in ["Provisional Votes"]:
```

Comment:
Not a big deal now, but thinking ahead to other states — would we maybe want to extract this array into a constant at the top of the file, like `NUM_VOTE_TYPES`?

Also, I don't know python conventions well enough, but would a set e.g. `{ "Provisional Votes" }` be a more efficient lookup here? I suppose it's irrelevant with a 1-length array, but I have wondered about the use of sets vs arrays with the `in` operator.

Reply:
You know, I suspect you are correct and that it is more efficient to use a set for this particular scenario! Gonna leave that for a follow up later though